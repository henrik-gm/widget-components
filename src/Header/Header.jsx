import React, { Component } from 'react';
import { coreLibrary, widgetModule } from 'kambi-widget-core-library';
import styles from './Header.scss';

class Header extends Component {

   /**
    * Constructs.
    * @param {object} props Header properties
    */
   constructor ( props ) {
      super(props);
      this.state = {
         hidden: this.props.hidden
      };
      this.toggleHeader = this.toggleHeader.bind(this);
   }

   /**
    * Called after mounting component
    */
   componentDidMount () {
      if ( this.state.hidden ) {
         // Collapse widget if needed by initial state
         widgetModule.setWidgetHeight(Header.HEIGHT);
      }
   }

   /**
    * Collapses or expands the widget
    * @returns {function|null} A callback baes on the action performed or null if no call back was provided
    */
   toggleHeader () {
      this.setState({ hidden: !this.state.hidden });
      if ( this.props.collapsable ) {
         if ( !this.state.hidden ) {
            widgetModule.setWidgetHeight(Header.HEIGHT);
            return ( this.props.onCollapse ) ? this.props.onCollapse() : null;
         }
         widgetModule.adaptWidgetHeight();
         return ( this.props.onExpand ) ? this.props.onExpand() : null;
      }
   }

   /**
    * Creates Header template.
    * @returns {XML}
    */
   render () {
      // Default classes to be added to all headers
      let cssClasses = 'l-flexbox l-pl-16 l-pr-16 KambiWidget-card-support-text-color';
      // If we have custom classes disregard default styling and load custom classes
      if ( this.props.className ) {
         cssClasses = this.props.className;
      } else {
         // Add classes depending on pageInfo
         cssClasses += (coreLibrary.pageInfo.pageType === 'home') ? ' l-pt-6 l-pb-6 ' + styles.kwHeader : ' KambiWidget-header l-pt-8 l-pb-8';
      }

      return (
         <header
            className={cssClasses}
            onClick={this.toggleHeader}>
            {this.props.children}
         </header>
      )
   }
}

Header.propTypes = {

   /**
    * Elements passed from parent
    */
   children: React.PropTypes.node.isRequired,

   /**
    * Sets header as collapsable
    */
   collapsable: React.PropTypes.bool,


   /**
    * Sets initial state
    */
   hidden: React.PropTypes.bool,

   /**
    * Callback after collapse
    */
   onCollapse: React.PropTypes.func,

   /**
    * Callback after expand
    */
   onExpand: React.PropTypes.func,

   /**
    * Custom styling.
    */
   className: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.bool])
};

Header.defaultProps = {
   className: false
};

/**
 * Holds the height of the header in pixels
 * @type {number}
 */
Header.HEIGHT = (coreLibrary.pageInfo.pageType === 'home') ? 37 : 40;

export default Header;
