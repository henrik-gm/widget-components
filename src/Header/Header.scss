.kwNotHome {
   font-size: 14px;
}
