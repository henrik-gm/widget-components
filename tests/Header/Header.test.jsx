import React, { Children } from 'react';
import Header from '../../src/Header/Header';
import ReactTestRenderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { widgetModule } from 'kambi-widget-core-library';

let mockPageType = 'home';

jest.mock('kambi-widget-core-library', () => ({
   widgetModule: {
      setWidgetHeight: jest.fn(),
      adaptWidgetHeight: jest.fn()
   },
   coreLibrary: {
      pageInfo: {
         get pageType() { return mockPageType; }
      }
   }
}));

describe('Header DOM rendering', () => {

   it('renders correctly with default props', () => {
      mockPageType = 'home';

      const tree = ReactTestRenderer.create(
         <Header>Test</Header>
      ).toJSON();

      expect(tree).toMatchSnapshot();
   });

   it('renders correctly with pageType != \'home\'', () => {
      mockPageType = 'not_home';

      const tree = ReactTestRenderer.create(
         <Header>Test</Header>
      ).toJSON();

      expect(tree).toMatchSnapshot();
   });

   it('renders correctly with custom classes', () => {
      mockPageType = 'home';

      const tree = ReactTestRenderer.create(
         <Header customClasses={'a b c'}>Test</Header>
      ).toJSON();

      expect(tree).toMatchSnapshot();
   });

});

describe('Header behaviour', () => {

   it('mounts in hidden state correctly', () => {
      widgetModule.setWidgetHeight.mockClear();

      mockPageType = 'home';

      const wrapper = mount(
         <Header hidden={true}>Test</Header>
      );

      expect(widgetModule.setWidgetHeight).toHaveBeenCalledTimes(1);
   });

   it('handles clicks correctly', () => {
      widgetModule.setWidgetHeight.mockClear();
      widgetModule.adaptWidgetHeight.mockClear();

      mockPageType = 'home';

      const wrapper = mount(
         <Header
            collapsable={true}
            customClasses='header-handle'
         >
            Test
         </Header>
      );

      wrapper.find('.header-handle').simulate('click');
      expect(widgetModule.setWidgetHeight).toHaveBeenCalledTimes(1);

      wrapper.find('.header-handle').simulate('click');
      expect(widgetModule.adaptWidgetHeight).toHaveBeenCalledTimes(1);
   });

   it('calls onExpand/onCollapse correctly', () => {
      widgetModule.setWidgetHeight.mockClear();
      widgetModule.adaptWidgetHeight.mockClear();

      mockPageType = 'home';

      const onExpandMock = jest.fn(),
         onCollapseMock = jest.fn();

      const wrapper = mount(
         <Header
            collapsable={true}
            onExpand={onExpandMock}
            onCollapse={onCollapseMock}
            customClasses='header-handle'
         >
            Test
         </Header>
      );

      wrapper.find('.header-handle').simulate('click');
      expect(onExpandMock).not.toHaveBeenCalled();
      expect(onCollapseMock).toHaveBeenCalledTimes(1);
      expect(widgetModule.setWidgetHeight).toHaveBeenCalledTimes(1);
      onExpandMock.mockClear();
      onCollapseMock.mockClear();

      wrapper.find('.header-handle').simulate('click');
      expect(onExpandMock).toHaveBeenCalledTimes(1);
      expect(onCollapseMock).not.toHaveBeenCalled();
      expect(widgetModule.adaptWidgetHeight).toHaveBeenCalledTimes(1);
   });

   it('ignores clicks when collapsable is not set', () => {
      mockPageType = 'home';

      const onExpandMock = jest.fn(),
         onCollapseMock = jest.fn();

      const wrapper = mount(
         <Header
            onExpand={onExpandMock}
            onCollapse={onCollapseMock}
            customClasses='header-handle'
         >
            Test
         </Header>
      );

      wrapper.find('.header-handle').simulate('click');
      expect(onExpandMock).not.toHaveBeenCalled();
      expect(onCollapseMock).not.toHaveBeenCalled();

      wrapper.find('.header-handle').simulate('click');
      expect(onExpandMock).not.toHaveBeenCalled();
      expect(onCollapseMock).not.toHaveBeenCalled();
   });
});
