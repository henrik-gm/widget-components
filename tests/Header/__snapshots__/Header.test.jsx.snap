// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header DOM rendering renders correctly with custom classes 1`] = `
<header
  className="a b c"
  onClick={[Function]}
>
  Test
</header>
`;

exports[`Header DOM rendering renders correctly with default props 1`] = `
<header
  className="l-flexbox l-pl-16 l-pr-16 KambiWidget-card-support-text-color KambiWidget-card-header-border l-pt-6 l-pb-6 "
  onClick={[Function]}
>
  Test
</header>
`;

exports[`Header DOM rendering renders correctly with pageType != 'home' 1`] = `
<header
  className="l-flexbox l-pl-16 l-pr-16 KambiWidget-card-support-text-color KambiWidget-card-header-border KambiWidget-header l-pt-8 l-pb-8 kwNotHome"
  onClick={[Function]}
>
  Test
</header>
`;
